File: 2305. Fair Distribution of Cookies/Solution.py
class Solution:
    def distributeCookies(self, cookies, k):
        self.answer = float('inf')
        self.counter = [0] * k
        self.findMaximum(cookies, 0, k)
        return self.answer
    
    def findMaximum(self, cookies, index, k):
        if index == len(cookies):
            max_val = 0
            for i in range(k):
                max_val = max(max_val, self.counter[i])
            self.answer = min(self.answer, max_val)
            return
        
        for i in range(k):
            self.counter[i] += cookies[index]
            self.findMaximum(cookies, index + 1, k)
            self.counter[i] -= cookies[index]
            if self.counter[i] == 0:
                break

File: 2305. Fair Distribution of Cookies/Solution.java
class Solution {
    int answer;
    int counter[];
    public int distributeCookies(int[] cookies, int k) {
        answer = Integer.MAX_VALUE;
        counter = new int[k];
        findMaximum(cookies, 0, k);
        return answer;
    }
    public void findMaximum(int[] cookies, int index, int k) {
        if (index == cookies.length) {
            int max = 0;
            for (int i = 0; i < k; i++) {
                max = Math.max(max, counter[i]);
            }
            answer = Math.min(answer, max);
            return;
        }
        for (int i = 0; i < k; i++) {
            counter[i] += cookies[index];
            findMaximum(cookies, index + 1, k);
            counter[i] -= cookies[index];
            if (counter[i] == 0) {
                break;
            }
        }
    }
}